### PR TITLE
Switch from sync to a tokio runtime

### PR DESCRIPTION
**Pull request by rbtcollins**
Being able to use async where relevant across the codebase makes it possible to use async constructs where relevant. rustup has many places where that can be useful: downloading channel metadata, distribution content, unpacking that content to disk, are work that could usefully proceed in parallel - and async provides a good abstraction for that.

We already have a sophisticated disk IO layer that accomodates various OS latency-inducing behaviours, and adapting that to async without any regressions could be very interesting too - but for now, it co-exists nicely with an async core.

One not necessarily obvious caveat - we can't use tokio::main because it doesn't provide the per-worker-thread injection that we use to facilitate isolated tests. (tl;dr std::env::var/args being global reflect reality well, but not writing tests that interact with those layers).

### CODE REVIEW

**Comment by djc** [comment]
Having briefly skimmed over some parts of this: what limits us from just using `#[tokio::main]` and making things natively async, rather than all the callback stuff and `block_on()` type calls?

**Comment by rbtcollins** [comment]
> Having briefly skimmed over some parts of this: what limits us from just using `#[tokio::main]` and making things natively async, rather than all the callback stuff and `block_on()` type calls?

This is in `draft` mode at the moment because it is incomplete.

The block-on stuff is scaffolding during the transition.

The starting point was:
sync main()
...
reqwest blocking http client
tokio runtime invoked here

With an async main, that will blow up at runtime, because you get
tokio runtime
async main()
...
reqwest blocking http client
tokio runtime invoked here - panic

If that is changed naively by just making main() async, and reqwest async, then we get rust's async function tainting everything and will change all the relevant calls to async, **but** there are a number of places where types need to change to be compatible with async requirements; so the commit becomes huge and unwieldy very quickly.

So the approach I'm taking is some scaffolding that lets us work this incrementally, and once complete it will just be nice async code.

**Comment by djc** [comment]
Makes sense, thanks for the explanation. Can you give an example of where/how types need to change? I worry that with this approach, you'll need another big set of commits to unwind all the `block_on()` stuff again after switching `main`. Instead, maybe it would be possible to use the top-down approach to isolate the required type changes in their own commits, and then jump directly to full async mode from there by adding `async` keywords all over the place?

**Comment by rbtcollins** [comment]
Once done the aggregate diff should be precisely the changes needed, and I can rebase squash it all away. We did this with the removal of `failure`. Many many small steps, nice outcome :)

**Comment by djc** [comment]
Oh, I guess that makes sense. Other projects I help maintain prefer to rebase-merge so that the small commits make it into main, which can be helpful if you need to chase down a regression later.

**Comment by rbtcollins** [comment]
Thats also valid to do; lets assess when this is at the next inflection point (which will be async native with no thunk, but no refactorings for better behaviour)

**Comment by rbtcollins** [comment]
This is nearly complete, I think its mostly mechanical now, except for the closures in the test suite - there's some lifetime challenges to address there

**Comment by rbtcollins** [comment]
This now complete modulo any refactoring of the commit series we might want to do.

The key notes are:
- we can't use tokio::main because it doesn't provide the per-worker-thread injection that we use to facilitate isolated tests. (tl;dr std::env::var/args being global reflect reality well, but not writing tests that interact with those layers).

File: Cargo.toml [hunk 1]
```diff
@@ -1,14 +1,14 @@
 [package]
-name = "rustup"
-version = "1.26.0"
-edition = "2021"
+build = "build.rs"
```
Comment (djc) [inline Cargo.toml]: I'd prefer not to have reorganization of `Cargo.toml` in this PR (for both crates), since this makes it much harder to see actual changes to the contents that are related to the main topic of this PR.

(Unless we move all of the actual `async` changes into one commit that can be reviewed independently, but actually it looks like these `Cargo.toml` changes are a step in the wrong direction IMO.)

File: Cargo.toml [hunk 2]
```diff
@@ -159,16 +162,24 @@ anyhow = "1.0.69"
 derivative = "2.2.0"
 enum_dispatch = "0.3.11"
 fs_at = "0.1.6"
+futures = "0.3"
+lazy_static = "1"
```
Comment (djc) [inline Cargo.toml]: AFAIK lazy_static has been deprecated in favor of once_cell (which we already have as a dependency).

Also it looks like lazy_static doesn't actually get used? I guess the `Cargo.toml` changes are mostly some rebase artifact?

File: download/src/lib.rs [hunk 3]
```diff
@@ -61,74 +61,81 @@ pub fn download_to_path_with_backend(
...
+                    0
+                };
 
-            possible_partial.seek(SeekFrom::End(0))?;
-
-            (possible_partial, downloaded_so_far)
-        } else {
-            (
-                OpenOptions::new()
+                // TODO: blocking call
```
Comment (djc) [inline download/src/lib.rs]: Should this be addressed before merging?

File: download/src/lib.rs [hunk 4]
```diff
@@ -61,74 +61,81 @@ pub fn download_to_path_with_backend(
...
+            };
 
-        file.borrow_mut()
-            .sync_data()
-            .context("unable to sync download to disk")?;
+            let file = RefCell::new(file);
 
-        Ok(())
-    }()
+            // TODO: the sync callback will stall the async runtime if IO calls block, which is OS dependent. Rearrange.
```
Comment (djc) [inline download/src/lib.rs]: Should this be addressed before merging?

File: src/bin/rustup-init.rs [hunk 5]
```diff
@@ -26,29 +26,36 @@ use rustup::cli::rustup_mode;
...
 
-    let process = OSProcess::default();
-    with(process.into(), || match maybe_trace_rustup() {
-        Err(e) => {
-            common::report_error(&e);
-            std::process::exit(1);
+    let process = OSProcess::new();
+    let mut builder = Builder::new_multi_thread();
+    builder.enable_all();
+    with_runtime(process.into(), builder, {
```
Comment (djc) [inline src/bin/rustup-init.rs:42]: Do we need the outer block here?

File: src/bin/rustup-init.rs [hunk 6]
```diff
@@ -26,29 +26,36 @@ use rustup::cli::rustup_mode;
...
 fn main() {
     #[cfg(windows)]
     pre_rustup_main_init();
 
-    let process = OSProcess::default();
-    with(process.into(), || match maybe_trace_rustup() {
-        Err(e) => {
-            common::report_error(&e);
-            std::process::exit(1);
+    let process = OSProcess::new();
```
Comment (djc) [inline src/bin/rustup-init.rs]: Why the change from `OSProcess::default()` to `OSProcess::new()` here? That seems unrelated.

File: src/cli/proxy_mode.rs [hunk 7]
```diff
@@ -12,7 +12,7 @@ use crate::{
 };
 
 #[cfg_attr(feature = "otel", tracing::instrument)]
-pub fn main(arg0: &str) -> Result<ExitCode> {
+pub async fn main(arg0: &str) -> Result<ExitCode> {
```
Comment (djc) [inline src/cli/proxy_mode.rs]: Have you checked the performance impact of these changes on proxy invocations?

File: src/cli/rustup_mode.rs [hunk 8]
```diff
@@ -84,8 +85,33 @@ where
     callee(cfg, matches)
 }
 
+async fn deprecated_async<'a, FF, F, B, R>(
+    instead: &str,
+    cfg: &'a mut Cfg,
+    matches: B,
+    callee: FF,
```
Comment (djc) [inline src/cli/rustup_mode.rs]: I would suggest writing this as `impl FnOnce(&'a mut Cfg, B) -> impl Future<Output = R> + 'a`, which seems easier to understand to me.

File: src/cli/rustup_mode.rs [hunk 9]
```diff
@@ -84,8 +85,33 @@ where
...
+    ));
+    (cfg.notify_handler)(Notification::PlainVerboseMessage(
+        "The exact API of this command may change without warning",
+    ));
+    (cfg.notify_handler)(Notification::PlainVerboseMessage(
+        "Eventually this command will be a true alias.  Until then:",
+    ));
+    (cfg.notify_handler)(Notification::PlainVerboseMessage(&format!(
+        "  Please use `rustup {instead}` instead"
+    )));
```
Comment (djc) [inline src/cli/rustup_mode.rs]: Given that this is exactly the same as in sync `deprecated()`, let's deduplicate? Alternatively, is there a good reason not to use the async variant for the `uninstall` command so we don't need both?

File: src/cli/self_update.rs [hunk 10]
```diff
@@ -356,9 +356,10 @@ fn canonical_cargo_home() -> Result<Cow<'static, str>> {
 }
 
 /// Installing is a simple matter of copying the running binary to
-/// `CARGO_HOME`/bin, hard-linking the various Rust tools to it,
-/// and adding `CARGO_HOME`/bin to PATH.
-pub(crate) fn install(
+/// `CARGO_HOME`/bin, hard-linking the various Rust tools to it, and adding
+/// `CARGO_HOME`/bin to PATH by dropping shell scripts around the place, except
+/// on windows, and also we need a linker on Windows...
```
Comment (djc) [inline src/cli/self_update.rs]: Seems unrelated, can we isolate this change/defer it for now?

File: src/cli/self_update.rs [hunk 11]
```diff
@@ -1074,10 +1080,10 @@ pub(crate) fn uninstall(no_prompt: bool) -> Result<utils::ExitCode> {
 /// (and on windows this process will not be running to do it),
 /// rustup-init is stored in `CARGO_HOME`/bin, and then deleted next
 /// time rustup runs.
-pub(crate) fn update(cfg: &Cfg) -> Result<utils::ExitCode> {
-    common::warn_if_host_is_emulated();
-
+pub(crate) async fn update(cfg: &Cfg) -> Result<utils::ExitCode> {
     use common::SelfUpdatePermission::*;
+
+    common::warn_if_host_is_emulated();
```
Comment (djc) [inline src/cli/self_update.rs:1085]: For a large change like this, I'd prefer not to make unrelated changes.

File: src/cli/setup_mode.rs [hunk 12]
```diff
@@ -1,5 +1,7 @@
 use anyhow::Result;
-use clap::{builder::PossibleValuesParser, value_parser, Arg, ArgAction, Command};
+use clap::{
+    builder::PossibleValuesParser, error::ErrorKind, value_parser, Arg, ArgAction, Command,
```
Comment (djc) [inline src/cli/setup_mode.rs]: This too seems unrelated.

File: src/currentprocess.rs [hunk 13]
```diff
@@ -139,20 +140,89 @@ pub fn with<F, R>(process: Process, f: F) -> R
...
-    PROCESS.with(|p| {
+/// Run a function in the context of a process definition and a tokio runtime.
+///
+/// The process state is injected into a thread-local in every work thread of
+/// the runtime, but this requires access to the runtime builder, so this
+/// function must be the one to create the runtime.
+pub fn with_runtime<'a, FR, R>(
+    process: Process,
+    mut runtime_builder: tokio::runtime::Builder,
+    fut: FR,
```
Comment (djc) [inline src/currentprocess.rs]: Suggest writing this as `impl Future<Output = R> + 'a`.

File: src/currentprocess.rs [hunk 14]
```diff
@@ -139,20 +140,89 @@ pub fn with<F, R>(process: Process, f: F) -> R
...
+) -> R
+where
+    FR: Future<Output = R> + 'a,
+{
+    ensure_hook();
+
+    let start_process = process.clone();
+    let unpark_process = process.clone();
+    let runtime = runtime_builder
+        // propogate to blocking threads
```
Comment (djc) [inline src/currentprocess.rs]: Typo: "propogate" -> "propagate" (also below).

**Review by djc — COMMENTED** [review]
This looks mostly okay to me.

File: src/bin/rustup-init.rs (hunk 5 above)
Comment (rami3l) [reply src/bin/rustup-init.rs:42]: @djc I believe this is just a `rustfmt` specificity.

**Review by rami3l — APPROVED** [review]

File: Cargo.toml (hunk 1 above)
Comment (rbtcollins) [reply Cargo.toml]: I can back that out, its a side effect of the vscode even better toml plugin. The actual dep changes of course are needed.

File: Cargo.toml (hunk 2 above)
Comment (rbtcollins) [reply Cargo.toml]: https://github.com/rust-lang/rustup/pull/3367#discussion_r1560042952

File: download/src/lib.rs (hunk 3 above)
Comment (rbtcollins) [reply download/src/lib.rs]: oh, I missed that I had these TODOs left. I will review.

File: src/cli/proxy_mode.rs (hunk 7 above)
Comment (rbtcollins) [reply src/cli/proxy_mode.rs]: Not rigorously, but I know that most of our proxy overhead is in manifest parsing today, so I doubt that we'd get a measurable change even if the runtime initialisation were a concern.

File: download/src/lib.rs (hunk 3 above)
Comment (rbtcollins) [reply download/src/lib.rs]: I think this is fine a as blocking call for now : we download a small number of files during install, and it is very unlikely that this is going to cause a performance problem. Eventually of course we should do it, but the goal here was to get a tokio runtime in place at parity, not to make perf/concurrency improvements, since those will require reasoning about resource limits and so on in a very careful fashion.

File: download/src/lib.rs (hunk 4 above)
Comment (rbtcollins) [reply download/src/lib.rs]: No. Same as above, we should do so eventually.

The cb passed in does two things:
 - write a block of the file, which can/will block
 - calls the next layer out callback, which will be our download rate tracker

As we only download one file at the moment, there is no performance concern about the IO blocking - a rearranged async callback would want to prevent more data being downloaded to avoid OOM if disk IO were to stall (e.g. on NFS); thus this is appropriate for now, and when we make it async, it will be in order of supporting concurrent download of files, which is a different feature that we can all bring together

File: src/bin/rustup-init.rs (hunk 6 above)
Comment (rbtcollins) [reply src/bin/rustup-init.rs]: It was the elevation of stderr_is_a_ttry into the OSProcess struct, that meant that either a manual Default implementation was needed, or new(), and I have no idea why I decided to do both :).

I made stderr_is_a_tty state to resurrect a previous efficiency optimisation we had for Windows, so that tty probing only happens one per process, not once per terminal construction. It really should be a separate patch; and will need redoing with your other refactoring.

File: src/bin/rustup-init.rs (hunk 5 above)
Comment (rbtcollins) [reply src/bin/rustup-init.rs:42]: rustfmt insisted on it.

File: src/cli/self_update.rs (hunk 11 above)
Comment (rbtcollins) [reply src/cli/self_update.rs:1085]: sorry, what unrelated change?

File: src/cli/setup_mode.rs (hunk 12 above)
Comment (rbtcollins) [reply src/cli/setup_mode.rs]: error::ErrorKind is moved from within main() to a file level import. I don't recall if I had to do that do deal with a transient state during the refactoring, or whether I could move it back now. That said, I usually prefer top of module imports, so we can break if out if you like, but I'm not super excited by it :)

File: src/currentprocess.rs (hunk 14 above)
Comment (rbtcollins) [reply src/currentprocess.rs]: fixed

File: Cargo.toml (hunk 2 above)
Comment (rbtcollins) [reply Cargo.toml]: lazy_static is because this branch started before 

commit 867dfc7d46e46bc718bb4b947291990bdf4c7e61
Author: Dirkjan Ochtman <dirkjan@ochtman.nl>
Date:   Mon Nov 6 13:41:22 2023 +0100

    Replace usage of lazy_static with once_cell


fixed.

File: src/currentprocess.rs (hunk 13 above)
Comment (rbtcollins) [reply src/currentprocess.rs]: done

File: src/bin/rustup-init.rs (hunk 6 above)
Comment (djc) [reply src/bin/rustup-init.rs]: I'm not sure I understand what you mean here. Seems that `stderr_is_a_tty` is already state in `OSProcess` on `master`, and `default()` and `new()` are interchangeable, so I'd prefer to remove this change from this PR.

File: src/cli/self_update.rs (hunk 11 above)
Comment (djc) [reply src/cli/self_update.rs:1085]: Changing the order of calling `warn_if_host_is_emulated()` vs importing from `SelfUpdatePermission`.

File: src/cli/setup_mode.rs (hunk 12 above)
Comment (djc) [reply src/cli/setup_mode.rs]: I prefer top of module imports too, but when reviewing a large change I prefer for these not to be intertwined with the other changes.

File: src/cli/self_update.rs (hunk 11 above)
Comment (rbtcollins) [reply src/cli/self_update.rs:1085]: ah yes. Do you want me to back it out?

Comment (djc) [reply src/cli/self_update.rs:1085]: Yes, I think that is preferrable.

**Review by djc — APPROVED** [review]

**Comment by djc** [comment]
Have rebased this on current master and addressed my own comments (pretty much all about unrelated changes). I think this is good to go. @rami3l do you want to take another look? I just went through and compared each commit to the original PR in its state before I touched it and the differences are pretty minimal.

File: src/cli/self_update.rs (hunk 11 above)
Comment (rami3l) [reply src/cli/self_update.rs:1085]: @djc Can we mark this as resolved now?

**Comment by rami3l** [comment]
Thanks a lot for your hard work, @rbtcollins and @djc!
